tracing/ops_tracing/buffer.py

```diff
@@ -0,0 +1,298 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+# file except in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+"""Buffer for tracing data in the ops-tracing extension."""
+
+from __future__ import annotations
+
+import contextlib
+import functools
+import logging
+import sqlite3
+from pathlib import Path
+from typing import Callable
+
+from typing_extensions import ParamSpec, TypeVar
```

Comment:
This one should be behind `if TYPE_CHECKING:`, because `typing_extensions` should only be a static type checking dependency, not a regular one.